seed/seed20.go

```diff
@@ -353,6 +425,93 @@ func (e *noSnapDeclarationError) Error() string {
 	return fmt.Sprintf("cannot find snap-declaration for snap name: %s", e.snapRef.SnapName())
 }
 
+type errorComponentNotInSeed struct {
+	error
+}
+
+func (s *seed20) lookupVerifiedComponent(cref naming.ComponentRef, snapRev snap.Revision, snapID, snapProvenance string, snapsDir string, handler ContainerHandler, tm timings.Measurer) (Component, error) {
+	snapName := cref.SnapName
+	compName := cref.ComponentName
+
+	resKey := ResourceKey{SnapID: snapID, Name: compName}
+	resRev, ok := s.resRevByResKey[resKey]
+	if !ok {
+		// No assertions might be ok if the component is optional, the
+		// caller should check for this error type in that case.
+		return Component{}, errorComponentNotInSeed{
+			fmt.Errorf("resource revision assertion not found for %s", compName)}
+	}
+	resPair, ok := s.resPairByResKey[resKey]
+	if !ok {
+		return Component{}, errorComponentNotInSeed{
+			fmt.Errorf("resource pair assertion not found for %s", compName)}
+	}
+
+	compPath := filepath.Join(s.systemDir, snapsDir,
+		fmt.Sprintf("%s_%d.comp", cref.String(), resRev.ResourceRevision()))
+	_, err := os.Stat(compPath)
+	if err != nil {
+		// error should be of type *PathError
+		return Component{}, errorComponentNotInSeed{err}
+	}
+
+	// Checks
+	cpi := snap.MinimalComponentContainerPlaceInfo(compName, snap.R(resRev.Revision()), snapName)
+	newPath, snapSHA3_384, resSize, err := handler.HandleAndDigestAssertedContainer(
+		cpi, compPath, tm)
+	if err != nil {
+		return Component{}, err
+	}
+	if newPath != "" {
+		compPath = newPath
+	}
+	if resRev.ResourceSize() != resSize {
+		return Component{}, fmt.Errorf(
+			"resource %s size does not match size in resource revision: %d != %d",
+			compName, resSize, resRev.ResourceSize())
+	}
+	if snapSHA3_384 != resRev.ResourceSHA3_384() {
+		return Component{}, fmt.Errorf(
+			"cannot validate resource %s, hash mismatch with snap-resource-revision",
+			compName)
+	}
+
+	if resRev.Revision() != resPair.Revision() {
+		return Component{}, fmt.Errorf(
+			"resource %s revision does not match revision in resource pair: %d != %d",
+			compName, resRev.Revision(), resPair.Revision())
+	}
+	if resPair.SnapRevision() != snapRev.N {
+		return Component{}, fmt.Errorf(
+			"resource %s pair revision does not match snap revision: %d != %d",
+			compName, resPair.SnapRevision(), snapRev.N)
+	}
+
+	if resRev.Provenance() != snapProvenance {
+		return Component{}, fmt.Errorf(
+			"resource revision provenance for %s does not match snap provenance: %s != %s",
+			compName, resRev.Provenance(), snapProvenance)
+	}
+	if resPair.Provenance() != snapProvenance {
+		return Component{}, fmt.Errorf(
+			"resource pair provenance for %s does not match snap provenance: %s != %s",
+			compName, resPair.Provenance(), snapProvenance)
+	}
```

Comment:
these checks could be done before looking at the file? is there a reason to do them after?